### ISSUE TITLE
[REVIEW]: FixedPointFinder: A Tensorflow toolbox for identifying and characterizing fixed points in recurrent neural networks

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#1003]
**Submitting author:** @mattgolub (<a href="http://orcid.org/0000-0003-4508-0537">Matthew Golub</a>)
**Repository:** <a href="https://github.com/mattgolub/fixed-point-finder" target ="_blank">https://github.com/mattgolub/fixed-point-finder</a>
**Version:** v1.0
**Editor:** @arokem
**Reviewer:** @malmaud
**Archive:** <a href="https://doi.org/10.5281/zenodo.1476209" target="_blank">10.5281/zenodo.1476209</a>

## Status

[![status](http://joss.theoj.org/papers/fcd2e48b36c44013cb7dfc177f9a3f04/status.svg)](http://joss.theoj.org/papers/fcd2e48b36c44013cb7dfc177f9a3f04)

Status badge code:

```
HTML: <a href="http://joss.theoj.org/papers/fcd2e48b36c44013cb7dfc177f9a3f04"><img src="http://joss.theoj.org/papers/fcd2e48b36c44013cb7dfc177f9a3f04/status.svg"></a>
Markdown: [![status](http://joss.theoj.org/papers/fcd2e48b36c44013cb7dfc177f9a3f04/status.svg)](http://joss.theoj.org/papers/fcd2e48b36c44013cb7dfc177f9a3f04)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/mattgolub/fixed-point-finder" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@malmaud, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.theoj.org/about#reviewer_guidelines. Any questions/concerns please let @arokem know.

✨ **Please try and complete your review in the next two weeks** ✨ 

### Conflict of interest

- [x] As the reviewer I confirm that I have read the [JOSS conflict of interest policy](https://github.com/openjournals/joss/blob/master/COI.md) and that there are no conflicts of interest for me to review this work.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/mattgolub/fixed-point-finder">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Version:** Does the release version given match the GitHub release (v1.0)?
- [x] **Authorship:** Has the submitting author (@mattgolub) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the function of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Authors:** Does the `paper.md` file include a list of authors with their affiliations?
- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **References:** Do all archival references that should have a DOI list one (e.g., papers, datasets, software)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @malmaud it looks like you're currently assigned as the reviewer for this paper :tada:.

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01003/joss.01003/10.21105.joss.01003.pdf)

**Comment by malmaud** [comment]
Hi @mattgolub, looking forward to reviewing this. Before digging deeper, you seem to be missing these items from the JOSS requirements:

* *Version: Does the release version given match the GitHub release (v1.0)?* - There is no tagged release on the project's Github page (https://github.com/mattgolub/fixed-point-finder/releases)
* *Automated tests: Are there automated tests or manual steps described so that the function of the software can be verified?* - I don't see any tests. 
* *Community guidelines: Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support* - I don't see any mention of guidelines in either the Readme or as a separate file.

**Comment by mattgolub** [comment]
Hi @malmaud, thank you for taking up this review and informing us of these missing items. Here is how we have worked to address these issues:

- *Version*: I've released v1.0. You can find the release [here](https://github.com/mattgolub/fixed-point-finder/releases/tag/v1.0).
- *Automated tests*: We had thought of the end-to-end example as a test and living documentation of the toolbox. Because you have brought this to our attention, we are also working to include a more formal testing suite that will run the FixedPointFinder and compare its results to ground truth results. This will be released by the end of the week, along with a corresponding update to the README.
- *Community guidelines*: I've added a separate file describing [contribution guidelines](https://github.com/mattgolub/fixed-point-finder/blob/master/CONTRIBUTING.md). We have added a reference to these guidelines at the end of the [README](https://github.com/mattgolub/fixed-point-finder/blob/master/README.md).

**Comment by mattgolub** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01003/joss.01003/10.21105.joss.01003.pdf)

**Comment by malmaud** [comment]
Great! I'll try it out on my own system soon.

**Comment by malmaud** [comment]
Everything seems good - the package installed and the included example runs successfully on my machine. It's clearly solving an important task relevant to many in the ML community. The article is clear and well-written. The codebase is small but non-trivial. The public API methods have good docstrings attached.

I think just a few things and it's ready for acceptance:

* As mentioned earlier, some kind of tests are needed. Doesn't have to be anything too fancy.
* The installation is a little complex. You have to clone this package, clone a package it depends on, manually add both to your Python path, manually look through the README to see the other Python packages it depends on, and then manually install those with pip.

I'd like to see if that can be simplified a bit - either this package can have a `setup.py` so it itself is installable with pip, or the packages it depends on could be listed in a requirements.txt file, or a Dockerfile could be included that bundles everything up, etc.

* It's Python 2 only at the moment. That's not a deal breaker necessarily, although it's unfortunate given the increasing use of Python 3 and I think it would take just a small amount amount of work to make it Python 3 compatible given the small amount of code involved (the package is on the order of 2000 lines of Python).

cf
* https://github.com/mattgolub/fixed-point-finder/issues/3
* https://github.com/mattgolub/fixed-point-finder/issues/2

**Comment by arokem** [comment]
I would strongly urge an update to Python 3. If you want others to use the software into the future, and combine it with other packages in the scientific Python eco-system that would be a good idea. Many major packages are going to stop supporting Python 2 as soon as just over a year from now (https://python3statement.org/). The publication of this paper seems like a good time to undertake this update.

**Comment by mattgolub** [comment]
@malmaud Thank you again for your time and effort in this review. I believe I have addressed the needs that you laid out for publication with JOSS:

-*Installation*: I've added requirements*.txt files along with corresponding instructions for a more streamlined installation in the README. I've noted this in the GitHub issue on this topic, and I've closed the issue.

-*Automated Testing*: I've added a testing suite along with corresponding instructions in the README.

-*Python 3*: The rationale for developing using Python 2.7 has been that a large set of collaborators (for whom this tool is immediately intended; including many at Google Brain/AI) are still primarily developing using Python 2.7. I see the very likely future need for compatibility with Python 3. I appreciate that you noted that this is "not a deal breaker." If it's alright with you, I'll take care of this in future work (i.e., after the JOSS paper is out).

**Comment by malmaud** [comment]
Thanks - all my objections have been addressed. The automated tests run to completion on my machine. 

I'm personally OK with adding Python 3 support sometime later, but I think that will ultimately be up to the editor.

@arokem, I recommend accepting at this time, contingent on you being OK with the Python 3 issue.

**Comment by arokem** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
```
Attempting PDF compilation. Reticulating splines etc...
```

**Comment by whedon** [comment]
[ :point_right: Check article proof :page_facing_up: :point_left: ](https://github.com/openjournals/joss-papers/blob/joss.01003/joss.01003/10.21105.joss.01003.pdf)

**Comment by arokem** [comment]
Sorry for the slowness here. 

@malmaud: thank you for the review! 👏 

The paper is ready to be accepted, but ust one more thing, @mattgolub : could you please add DOIs for all the references in your bibliography? Once that's done, could you please create an archive for the software as it is after making all these changes (e.g., by tagging a release and uploading to zenodo.org). Once we have that, we can go ahead with the acceptance and archiving of the paper.

**Comment by mattgolub** [comment]
@arokem As requested, I've added DOIs to the bibliography and created a Zenodo release (10.5281/zenodo.1476209). Thanks again for all of your efforts here. 

@malmaud Thanks again for the review.

**Comment by arokem** [comment]
@whedon set 10.5281/zenodo.1476209 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.1476209" target="_blank">10.5281/zenodo.1476209</a> is the archive.

**Comment by arokem** [comment]
@arfon : I believe this article is ready to go!

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/44

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/44, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by arfon** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/45
1. Wait a couple of minutes to verify that the paper DOI resolves [https://doi.org/10.21105/joss.01003](https://doi.org/10.21105/joss.01003)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? notify your editorial technical team...

**Comment by arfon** [comment]
@malmaud - many thanks for your review here and to @arokem for editing this submission ✨ 

@mattgolub - your paper is now accepted into JOSS and your DOI is https://doi.org/10.21105/joss.01003 :zap: :rocket: :boom:

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](http://joss.theoj.org/papers/10.21105/joss.01003/status.svg)](https://doi.org/10.21105/joss.01003)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.01003">
  <img src="http://joss.theoj.org/papers/10.21105/joss.01003/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: http://joss.theoj.org/papers/10.21105/joss.01003/status.svg
   :target: https://doi.org/10.21105/joss.01003
```
This is how it will look in your documentation:

[![DOI](http://joss.theoj.org/papers/10.21105/joss.01003/status.svg)](https://doi.org/10.21105/joss.01003)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: http://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://www.flipcause.com/secure/cause_pdetails/Mjk3Nzk=
